2793 - Add validation indicator to IdsTab

anhallbe opened the pull request:
**Explain the details for making this change. What existing problem does the pull request solve?**
<!--
Example: When "Adding a function to do X",
explain why it is necessary to have a way to do X.
-->
Adds an error indicator to IdsTab when its' IdsTabContent contains an invalid input. 

**Related github/jira issue (required)**: Fixes #2793

**Summary:**
- Adds an `error` slot to IdsTab for customization (custom icon, tooltips, legacy IDS compat etc)
- Adds a `has-error` attribute to IdsTab which can be used to display the slotted `error`.
- Show the error in the tab title when there is a validation error inside the tab content.
- The IdsValidateMixin's `validate` event now bubbles, and has stronger typings.

**Related fixes:**
- Adds missing imports in IdsTab (icon & dismiss button)
- Fix vertical alignment of dismiss button

<img width="1087" alt="image" src="https://github.com/user-attachments/assets/b65be87d-fa6b-46e0-9728-799416dd2905">


**Steps necessary to review your pull request (required)**:

- Check out the branch
- `npm start`
- Go to http://localhost:4300/ids-tabs/validation.html
- Go to every test/sample page that contains either form validation or tabs, and make sure they still work.

**Included in this Pull Request**:
- [x] Some documentation for the feature.
- [x] A test for the bug or feature.
- [x] A note to the change log.

anhallbe commented on src/components/ids-tabs/ids-tabs-context.ts:
```diff
@@ -54,6 +55,20 @@ export default class IdsTabsContext extends Base {
+    this.onEvent('validate', this, () => {
+      this.querySelectorAll<IdsTabContent>('ids-tab-content').forEach((content) => {
+        // TODO: Figure out how to extract the type from the mixin
+        type ValidationMixinElement = HTMLElement & { isValid: boolean };
```
This

tmcconechy replied:
I would include the type in the event data so that you can do this...

```
this.onEvent('validate', this, (e: Event) => {
```

Somewhere around here but you changed it https://github.com/infor-design/enterprise-wc/blob/main/src/mixins/ids-validation-mixin/ids-validation-mixin.ts#L202 ? 

Maybe the whole rule including the type?
```
{ detail: { elem: this, value: (this as IdsInputInterface).value, isValid, rule: thisRule.rule } 
```

anhallbe replied:
Added `IdsValidateEvent` and `IdsValidatedElement` in [bac6645](https://github.com/infor-design/enterprise-wc/pull/2801/commits/bac664593a570d5fb1664c8d73fc2be2fa506a7a)

anhallbe commented on src/components/ids-tabs/ids-tabs-context.ts:
```diff
@@ -54,6 +55,20 @@ export default class IdsTabsContext extends Base {
+        // TODO: Figure out how to extract the type from the mixin
+        type ValidationMixinElement = HTMLElement & { isValid: boolean };
+
+        const tab = this.tabList?.querySelector<IdsTab>(`ids-tab[value="${content.value}"]`);
```
Maybe a method `IdsTabs.getTab(value: string): IdsTab` would make sense?

tmcconechy replied:
Or a getter `get tabByValue` in case we need id / ect later.

anhallbe replied:
Added a `tabs` getter in [bac6645](https://github.com/infor-design/enterprise-wc/pull/2801/commits/bac664593a570d5fb1664c8d73fc2be2fa506a7a)

anhallbe commented on src/components/ids-tabs/ids-tabs-context.ts:
```diff
@@ -54,6 +55,20 @@ export default class IdsTabsContext extends Base {
+
+        const tab = this.tabList?.querySelector<IdsTab>(`ids-tab[value="${content.value}"]`);
+        if (tab) {
+          const validated = content.querySelectorAll<ValidationMixinElement>('[validate]');
```
Not sure if all `[validate]` elements are guaranteed to be `ValidationMixinElement`. May need some form of `instanceof` check

tmcconechy replied:
Yes they "should" all use the mixin for any components that support it. See also ids-form.ts

anhallbe replied:
Added `isIdsValidatedElement(element)` in 9d00dd260111cba8f15bc9e802994147ad0d4b2f

anhallbe commented on src/mixins/ids-validation-mixin/ids-validation-mixin.ts:
```diff
@@ -199,7 +200,9 @@ const IdsValidationMixin = <T extends Constraints>(superclass: T) => class exten
       this.isTypeNotValid = null;
-      this.triggerEvent('validate', this, { detail: { elem: this, value: (this as IdsInputInterface).value, isValid } });
+      if (isValid !== wasValid) { // Only trigger if the state has changed, for performance.
+        this.#triggerValidateEvent(isValid);
```
This used to fire on every keypress. Added the `isValid !== wasValid` check for performance, but I'm not sure if this breaks anything.

tmcconechy replied:
This seems ok. But does it account for the case of clearing the field? i.e. reseting the state by clearing the text ect.

anhallbe replied:
Like this?

https://github.com/user-attachments/assets/967e1b9b-f6ff-4bda-9c3c-d05c4018d921

anhallbe commented on src/mixins/ids-validation-mixin/ids-validation-mixin.ts:
```diff
@@ -563,6 +566,17 @@ const IdsValidationMixin = <T extends Constraints>(superclass: T) => class exten
+        value: (this as IdsInputInterface).value,
+        isValid,
+      },
+      bubbles: true,
```
Note: Added `bubbles: true` here. Not sure if that has any unintended consequences?

tmcconechy replied:
It will just bubble if you do that. Its probably what we want here 👍🏻

tmcconechy commented:
> Handle different validation error levels (error, warning, info)

Be nice if you did... Not required for me

  > Add an indicator to the "more" menu in case the invalid tab is overflowed

We didnt have this but i think if we did it would just be the icon next to the item.

  >  Add an option to disable the error indicator?
 
 Yes maybe DISABLE_VALIDATION. What was the `VALIDATION_HAS_ERROR` for? Just a flag?
 
 > Handle / test validation errors from ids-enterprise controls?
 
 Not sure?
 
  > Right now this is implemented using the validation-has-error attribute. Not sure if this is the intended purpose of the attribute?
 
 I dont know what that is for? Its only used on radio button. I think it can go. I dont see what its doing maybe just to add the class and show it has an error? Would use hasError instead maybe 
  
> The 'validated' event now bubbles so that parent elements can detect it (IdsTab, IdsModal, IdsForm etc). Not sure if there's a better/preferred way to do this?
I think this one makes sense to do it that way.

> Test with more tab variants (module, counts etc).
Can push to later...

> Needs docs / changeling

Please and maybe a test.

anhallbe commented:
Needs a bit of work to align the icon for the `dismissible`, `count` and `module` cases (and all of these combined with `orientation="vertical"`...)

<img width="1840" alt="image" src="https://github.com/user-attachments/assets/405aadc2-b626-41b7-abfa-21fcd2cb9964">

ericangeles commented:
@anhallbe, have we considered the RTL style here?

ericangeles commented on tests/ids-tabs/ids-tabs.spec.ts:
```diff
@@ -196,4 +197,36 @@ test.describe('IdsTabs tests', () => {
+        const input = content.locator('ids-input').first();
+        const icon = tab.locator('ids-icon[icon="error"]');
+
+        await input.evaluate((el: IdsInput) => { el.value = 'hello'; });
```
Maybe we can avoid repetition here by creating a function?

```js
const setInputValue = async (value: string) => {
  await input.evaluate((el: IdsInput) => { el.value = value; });
};
```

ericangeles commented on src/components/ids-tabs/ids-tabs-context.ts:
```diff
@@ -54,6 +55,20 @@ export default class IdsTabsContext extends Base {
       this.#changeContentPane(this.value, this.value);
     });
 
+    this.onEvent('validate', this, (e: IdsValidateEvent) => {
```
This looks good to me. 👍🏼

anhallbe commented:
> @anhallbe, have we considered the RTL style here?

anhallbe commented:
> > Handle different validation error levels (error, warning, info)
> 
> Be nice if you did... Not required for me

Not implemented right now. But the icon is now slotted so people can implement that if they need to: cf59264

> > Add an indicator to the "more" menu in case the invalid tab is overflowed
> 
> We didnt have this but i think if we did it would just be the icon next to the item.

I guess for feature parity this can be saved for later, then?

> > Add an option to disable the error indicator?
> 
> Yes maybe DISABLE_VALIDATION. What was the `VALIDATION_HAS_ERROR` for? Just a flag?

Added a slot for the error, which can be hidden: cf59264

> > Handle / test validation errors from ids-enterprise controls?
> 
> Not sure?

Added a slot for the error, so this can be implemented by consumers in the meantime: cf59264

> > Right now this is implemented using the validation-has-error attribute. Not sure if this is the intended purpose of the attribute?
> 
> I dont know what that is for? It's only used on radio button. I think it can go. I dont see what it's doing maybe just to add the class and show it has an error? Would use hasError instead maybe

Switched to `has-error`.


> > Test with more tab variants (module, counts etc).
> > Can push to later...

@tmcconechy there appear to be an existing issue with the "dismissible" button (and this new error icon) being misaligned.

<img width="528" alt="image" src="https://github.com/user-attachments/assets/e801e36a-2aa7-4acb-adfc-7f4ef74c5e25">


> > Needs docs / changeling
> 
> Please and maybe a test.

Added docs & test

tmcconechy commented:
@anhallbe usually in that case just go ahead and include i fix with it (when i see issues in the same place I just correct them)

anhallbe commented:
The alignment issue has been fixed, and I added a note in the Changelog (v 1.5.0). I think this is ready for review/test/merge.

tmcconechy commented:
@anhallbe This is great! Can you just check the tests https://github.com/infor-design/enterprise-wc/actions/runs/10826090266/job/30044182690?pr=2801#step:12:2667

Yep, This will be in 1.6.0 or 1.5.1 depending on semantic versioning as things change over the sprint. (Probably 1.6.0 considering whats in the Queue)

anhallbe commented:
@tmcconechy the failing test was because of my change in when the `validate` event fired.

I had a change of heart regarding that one so I reverted the change in 41533ba. The `'validate'` event will again always fire when `checkValidation()` is called. This is consistent with how native elements always fire the `invalid` event when `checkValidity()` is called.

Also added a 1.6.0 section to the Changelog
